exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,262 @@
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	var aggError error
+	checkpointSet.ForEach(func(record export.Record) {
+		agg := record.Aggregator()
+		if sum, ok := agg.(aggregator.Sum); ok {
+			value, err := sum.Sum()
+			if err != nil {
+				aggError = err
+				// TODO: halp, where are we doing here?
+				return
+			}
+
+			c, err := e.getCounter(record)
+			if err != nil {
+				// TODO: log a warning here?
+				return
+			}
+
+			c.Add(value.AsFloat64())
+		}
+
+		if gauge, ok := agg.(aggregator.LastValue); ok {
+			lv, _, err := gauge.LastValue()
+			if err != nil {
+				aggError = err
+				// TODO: halp
+				return
+			}
+
+			g, err := e.getGauge(record)
+			if err != nil {
+				// TODO: log a warning here?
+				return
+			}
+			g.Set(lv.AsFloat64())
+		}
+	})
+
+	return aggError
+}
+
+func (e *Exporter) getCounter(record export.Record) (prometheus.Counter, error) {
+	e.Lock()
+	defer e.Unlock()
+
+	desc := record.Descriptor()
+	if c, ok := e.counters[desc]; ok {
+		return c, nil
+	}
+
+	counterVec, err := e.getCounterVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	counter := counterVec.With(labelsToTags(record.Labels()))
+
+	e.counters[desc] = counter
+	return counter, nil
+}
+
+func (e *Exporter) getGauge(record export.Record) (prometheus.Gauge, error) {
+	e.Lock()
+	defer e.Unlock()
+
+	desc := record.Descriptor()
+	if g, ok := e.gauges[desc]; ok {
+		return g, nil
+	}
+
+	gaugeVec, err := e.getGaugeVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	gauge := gaugeVec.With(labelsToTags(record.Labels()))
+
+	e.gauges[desc] = gauge
+	return gauge, nil
+}
+
+func (e *Exporter) getCounterVec(desc *export.Descriptor, labels export.Labels) (*prometheus.CounterVec, error) {
+	id, tagKeys := getCanonicalID(desc, labels)
+
+	if c, ok := e.counterVecs[id]; ok {
+		return c, nil
+	}
+
+	c := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: Sanitize(desc.Name()),
+			Help: desc.Description(),
+		},
+		tagKeys,
+	)
+
+	if err := e.registerer.Register(c); err != nil {
+		return nil, err
+	}
+
+	e.counterVecs[id] = c
+	return c, nil
+}
+
+func (e *Exporter) getGaugeVec(desc *export.Descriptor, labels export.Labels) (*prometheus.GaugeVec, error) {
+	id, tagKeys := getCanonicalID(desc, labels)
+
+	if g, ok := e.gaugeVecs[id]; ok {
+		return g, nil
+	}
+
+	g := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: Sanitize(desc.Name()),
+			Help: desc.Description(),
+		},
+		tagKeys,
+	)
+
+	if err := e.registerer.Register(g); err != nil {
+		return nil, err
+	}
+
+	e.gaugeVecs[id] = g
+	return g, nil
+}
+
+func getTagKeys(keys []core.KeyValue) []string {
+	tagKeys := make([]string, 0, len(keys))
+	for _, kv := range keys {
+		tagKeys = append(tagKeys, Sanitize(string(kv.Key)))
+	}
+	return tagKeys
+}
+
+func getCanonicalID(desc *export.Descriptor, labels export.Labels) (string, []string) {
```

Comment:
The way this is written, we **need** to use the defaultkeys Batcher, which ensures that the labels on each export Record are the same.   I think this is another reason to integrate the exporter with the controller (and let it construct a Batcher).

The defaultkeys Batcher will let us simplify this code, since we can be fairly sure that the calls to `With()` will match the tags declared for the vector.

Reply:
This method is not needed anymore so I have just deleted. 

Do you mean this because of the `labels.Encoded()` call? because I've just moved it to another place. 

I guess I should make the Prometheus exporter implements the `LabelEncoder` with a defaultkeys batcher for the moment?

Reply:
My understanding is that no two -Vec types can exist with the same name and different key sets. The defaultkeys batcher will ensure that the tag sets are always equal--that's why it makes sense as a default. I still envision other batchers being created to support greater configuration, but for now yes, I would use defaultkeys.

Also, yes, it makes sense to use the LabelEncoder interface to compute anything needed to map the -Vec type into the handle-like interface. If the `map[string]string` input to prometheus can be changed to a slice of values (as you said it could), then maybe the `Encode()` function will compute a slice of values.